### PR TITLE
Che Operator updates DWO and DWCO resources

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
- Che Operator sets annotations DWO and DWCO resources
- Che Operator syncs DWO and DWCO resources by updating them if hashes are different 
- Add DWO and DWCO tests for OCP

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

Che Operator does not update DWO and DWCO resources https://github.com/eclipse/che/issues/19664

### How to test this PR?

1. Run OCP cluster
2. Update ./local-debug.sh [1] by setting some old version of DWO and DWCO, for instance 
```
DEV_WORKSPACE_CONTROLLER_VERSION="v0.2.1"
DEV_WORKSPACE_CHE_OPERATOR_VERSION="v7.28.0"
```
3. Update CR [2] to turn off OAuth (spec.auth.openShiftoAuth: false)
4. Update CR [2] to turn on DevWorkspace (spec.devWorkspace.enable: true)
5. Run `./local-debug.sh`, wait until everything is deployed
6. Stop `./loca-debug.sh` and update DWO and DWCO, for instance:
```
DEV_WORKSPACE_CONTROLLER_VERSION="main"
DEV_WORKSPACE_CHE_OPERATOR_VERSION="main"
```
7. Run `./local-debug.sh`, wait until DWO and DWCO resources are updated. You can see that in the logs.
8. Check `devworkspace-che` and `devworkspace-controller` workspaces that all pods are running
9. Create workspaces and check that all pods are running.
```
oc apply -f https://raw.githubusercontent.com/che-incubator/devworkspace-che-operator/main/samples/flattened_theia-nodejs.yaml -n eclipse-che
oc apply -f https://raw.githubusercontent.com/devfile/devworkspace-operator/main/samples/flattened_theia-next.yaml -n eclipse-che
```

[1]  https://github.com/eclipse-che/che-operator/blob/main/local-debug.sh#L21-L22
[2] https://github.com/eclipse-che/che-operator/blob/main/deploy/crds/org_v1_che_cr.yaml

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [Nightly OLM bundle is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-nightly-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by sleshchenko — COMMENTED** [review]
It's a draft so technically I should not approve it )

The idea with hashes and changes LGTM
So, we're getting updates for DevWorkspace and DevWorkspace Che Operators with pretty minimal effort without diving deep which objects each of the operators brings.

Only one question: will syncObject only try to update objects, like deployment?
Asking since fields like selector is immutable, and if labels and selector is changes, deployment needs to be removed and recreated.

If Che Operator won't try to recreate - that's fine but Platform and Controller teams should be aware of it and do not touch immutable fields.

**Review by AndrienkoAleksandr — APPROVED** [review]

**Review by flacatus — APPROVED** [review]

File: pkg/deploy/test_util.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+//
+// Copyright (c) 2012-2019 Red Hat, Inc.
```
Comment (nickboldt) [inline pkg/deploy/test_util.go:2]: new file, who dis? (please fix copyright date to current year)

**Comment by nickboldt** [comment]
```
DEV_WORKSPACE_CONTROLLER_VERSION="v0.2.1"
DEV_WORKSPACE_CHE_OPERATOR_VERSION="v7.28.0"
```

Will this logic work with CRW versions too? eg., an update from DWO/DWCO 2.8 -> 2.9 ?

**Review by mmorhun — APPROVED** [review]
LGTM

**Pull request merged**